dfastmi/io/RiversObject.py

```diff
@@ -0,0 +1,344 @@
+    prop_c: List[List[Tuple[float, ...]]]
+    proprate_high: List[List[float]] # only for version = 1
+    proprate_low: List[List[float]] # only for version = 1
+    qbankfull: List[List[float]] # only for version = 1
+    qmin: List[List[float]] # only for version = 1
+    qlevels: List[List[Tuple[float, float, float, float]]] # only for version = 1
+    dq: List[List[Tuple[float, float]]] # only for version = 1
+    
+    _river_data : RiverParameterData
+
+
+    def __init__(self, filename: str = "rivers.ini"):
+        self._read_rivers_file(filename)
+
+    def _read_rivers_file(self, filename):
+        """
+        Read a configuration file containing the river data.
+
+        Read the configuration file containing the listing of various branches/reaches
+        and their associated default parameter settings.
+
+        Parameters
+        ----------
+        filename : str
+            The name of the river configuration file (default "rivers.ini").
+
+        """
+        config: configparser.ConfigParser
+        
+        # read the file
+        config = configparser.ConfigParser()
+        with open(filename, "r") as configfile:
+            config.read_file(configfile)
+
+        self._river_data = RiverParameterData(config)
+
+        # initialize rivers dictionary
+        iversion = self._validate_version_in_file(filename, config)
+
+        self._verify_checksum_rivers(config, filename)
+        
+        # parse branches
+        self._parse_branches(config)
+
+        # parse reaches and discharge locations
+        self._parse_reaches_and_discharge_locations(config)
+
+        # collect the values for all other quantities
+        self.nreaches = [len(x) for x in self.allreaches]
+
+        # initialize river data object
+        self._river_data.initialize(self.branches, self.nreaches)
+        
+        # call the specific reader for the file version
+        if iversion == 1:
+            self._read_rivers_legacy()
+            
+        else: # iversion == 2
+            self._read_rivers() 
+
+    def _parse_reaches_and_discharge_locations(self, config):
+        self.allreaches = []
+        self.qlocations = []
+        for branch in self.branches:
+            qlocation = config[branch]["QLocation"]
+            self.qlocations.extend([qlocation])
+
+            i = 0
+            reaches = []
+            while True:
+                i = i + 1
+                try:
+                    reach = config[branch]["Reach" + str(i)]
+                    reaches.extend([reach])
+                except:
+                    break
+            self.allreaches.extend([reaches])
+
+    def _parse_branches(self, config):
+        self.branches = [k for k in config.keys()]
+        self.branches.remove("DEFAULT")
+        self.branches.remove("General")
+
+    def _validate_version_in_file(self, filename, config):
+        try:
+            file_version = config["General"]["Version"]
```

Comment:
I cry myself to sleep at night because of such statements... :sob:

We can make use of the .get() in a similar way as mentioned before, for multidimensionals you just need to use it twice...

https://www.geeksforgeeks.org/how-to-use-dict-get-with-multidimensional-dict/

Reply:
Refactoring & improvements should be done in separate issue for each component

Reply:
https://issuetracker.deltares.nl/browse/DFAST-113